plugin-scalebar/src/main/java/com/mapbox/maps/plugin/scalebar/ScaleBarImpl.kt

```diff
@@ -300,81 +326,156 @@ class ScaleBarImpl : ScaleBar, View {
     }
   }
 
-  private fun drawText(canvas: Canvas, text: String, x: Float, y: Float, lastUnitBarText: Boolean) {
-    var safeX = x
-
-    // As an optimization only check if it goes beyond right view for the last unit bar text
-    if (lastUnitBarText) {
-      // Check if it goes beyond the right margin of the view
-      val textWidthPx = textPaint.measureText(text)
-      val textMaxRightPx = x + strokePaint.strokeWidth + when (textPaint.textAlign) {
-        Paint.Align.LEFT -> textWidthPx
-        Paint.Align.CENTER -> textWidthPx / 2
-        Paint.Align.RIGHT, null -> 0F
-      }
-      if (textMaxRightPx > width) {
-        // Move it away from right margin enough to fit
-        safeX -= (textMaxRightPx - width)
-      }
-    }
-
-    // Check if it goes beyond the left margin of the view
-    if (safeX - (strokePaint.strokeWidth / 2) < 0) {
-      safeX += (strokePaint.strokeWidth / 2)
-    }
-
+  private fun drawText(
+    canvas: Canvas,
+    text: String,
+    x: Float,
+    y: Float,
+  ) {
     if (settings.showTextBorder) {
-      canvas.drawText(text, safeX, y, strokePaint)
+      canvas.drawText(text, x, y, strokePaint)
     }
-    canvas.drawText(text, safeX, y, textPaint)
+    canvas.drawText(text, x, y, textPaint)
   }
 
   /**
-   * @return A tuple with the scale bar segment distance and the amount of bar segments
+   * @return a new configuration holding the values to be able to render the scale bar
    */
   @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
-  internal fun findSuitableScaleBarSegments(maxDistance: Float): Pair<Float, Int> {
-    var pair: Pair<Int, Int> = scaleTable[0]
-    for (i in 1 until scaleTable.size) {
-      pair = scaleTable[i]
-      val distance = pair.first
-      if (distance > maxDistance) {
-        // use the last scale here, otherwise the scale will be too large
-        pair = scaleTable[i - 1]
-        break
-      }
-    }
-    val distance = pair.first.toFloat()
-    var rectCount = pair.second
-    var unitDistance = distance / rectCount
+  internal fun calculateSegmentsConfiguration(
+    maxDistance: Float,
+    distancePerPixel: Float,
+    scaleTable: List<Pair<Int, Int>>,
+    textPaint: Paint,
+    strokeWidth: Float,
+    unit: String,
+    rightMarginPx: Int
+  ): SegmentsConfiguration {
+    // Find the entry where the distance just fits in the given `maxDistance`
+    val pair =
+      scaleTable.lastOrNull { (distance: Int, _: Int) -> distance <= maxDistance } ?: scaleTable[0]
+
+    var distance: Float = pair.first.toFloat()
+    var rectCount: Int = pair.second
+    var unitDistance: Float = distance / rectCount
+
     // When maxDistance is small (i.e. high zoom levels near the poles) then
     // the `distance` might be bigger than maxDistance. This loop will keep removing
     // bar divisions (rectCount) until it fits
     while (unitDistance * rectCount > maxDistance && rectCount > 0) {
       rectCount--
+      distance = unitDistance * rectCount
```

Comment:
Maybe move out of the loop below it?

Reply:
I'm not sure I follow, given the following scenario:
- `unitDistance * rectCount` is less than `maxDistance`
- `rectCount` is > 0
wouldn't that change the logic if I take it out or in the else?

The value could still be the same (given that `unitDistance` is `distance / rectCount`) but there could also be float conversion issues.

Reply:
Indeed, I didn't take into account that loop may not run at all.